### PR TITLE
week 5 assessment

### PR DESCRIPTION
**Pull request by sanjuel**
# Submitting Week 5 Assessment

### ✅ Completed Files

Mark the following as confirmation that the files in this repository are complete.

- [x] JavaScript Coding Challenges
- [x] Ruby Coding Challenges
- [x] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:
This one was for sure my weakest assessment. Gonna try and work through the problems again with some more time later.

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -21,8 +21,53 @@ const secretCodeWord2 = "Gobbledygook"
 const secretCodeWord3 = "Eccentric"
 // Expected output: "3cc3ntr1c"
 
+describe("code", ()=>{
+    it("takes in a string and returns a coded message. The coded message converts 'a' to 4, 'e' to 3, 'i' to 1, and 'o' to 0.", ()=>{
+        expect(code(secretCodeWord1)).toEqual("L4ck4d41s1c4l")
+        expect(code(secretCodeWord2)).toEqual("G0bbl3dyg00k")
+        expect(code(secretCodeWord3)).toEqual("3cc3ntr1c")
+    })
+})
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:30]: Be sure to move those variables into the test!
```javascript
describe(“codedMessage”, ()=> {
    it(“takes in a string and returns a coded message”, () => {
      const secretCodeWord1 = “Lackadaisical”
      const secretCodeWord2 = “Gobbledygook”
      const secretCodeWord3 = “Eccentric”
      expect(code(secretCodeWord1)).toEqual(“L4ck4d41s1c4l”)
      expect(code(secretCodeWord2)).toEqual(“G0bbl3dyg00k”)
      expect(code(secretCodeWord3)).toEqual(“3cc3ntr1c”)
    })
})

File: code-challenges.test.js [hunk 2]
```diff
@@ -21,8 +21,53 @@ const secretCodeWord2 = "Gobbledygook"
...
+
+
 // b) Create the function that makes the test pass.
 
+//Pseudo Code
+//Create a function named code
+// function takes in a string
+//create a variable to put new string into
+//create logic that pushes numbers instead of strings when the number is equal to A, E, I, or O.
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:39]: What happened to the capitals?

File: code-challenges.test.js [hunk 3]
```diff
@@ -21,8 +21,53 @@ const secretCodeWord2 = "Gobbledygook"
...
+        }
+    }
+    return newString
+}
+
+// code(secretCodeWord1)
+// code(secretCodeWord2)
+// code(secretCodeWord3)
+
+//I'm getting the error that newString is not a function. I feel like doing backend data stuff for a week has eroded my javscript skills :(. I will try to keep working on this problem later but for now that is my blocker.
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:69]: Correct! Your function is named code not newString!

File: code-challenges.test.js [hunk 4]
```diff
@@ -21,8 +21,53 @@ const secretCodeWord2 = "Gobbledygook"
...
+            newString[i]("3")
+    }
+        else if(string[i] === 'i') {
+            newString[i]("1")
+}
+        else if(string[i] === 'i') {
+            newString[i]("1")
+        }
+    }
+    return newString
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:62]: This is 90% good. Just missing some edge cases of capital letters and you are all set.

File: code-challenges.test.js [hunk 5]
```diff
@@ -34,8 +79,35 @@ const letterA = "a"
...
+const arrayFromLetter = (array, letter) => {
+    return letterArray = array.filter((value) => {
+        return value.toLowerCase.includes(letter)
+    })
+}
+
+arrayFromLetter(letterA)
+arrayFromLetter(letterE)
+
+//I'm now getting the same error, that .filter is not a function. I know i ran into this problem before, and knew how to fix it, but I can't remember the steps to take here. I've always had problems with getting higher order functions to work, so I need to work on them more.
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:109]: Your test is only providing one variable but your function expects two. Looks like your logic and psuedo code are perfectly sound it's just the set up details getting past you. I know that feeling.

File: code-challenges.test.js [hunk 6]
```diff
@@ -50,3 +122,32 @@ const hand4 = [7, 2, 7, 2, 7]
...
+const fullHouse = (array) => {
+    let newHand = []
+ for(let i = 0; i < newHand.length; i++) {
+    if (array[i] === array[i-1] && array[i] === array[i +1])
+    return newHand.push(array)
+    else if (array[i] === array[i-2] && array[i] === array[i +2])
+ } return newHand.push(array)
+}
+
+// I tried to work through this one as if it was an interview question. I don't think id' have gotten the job. My blocker is that I don't know how to do this without having a million combinations that I'd have to code for. I feel like I can do everything separately, but when I put them all together my mind goes blank. I need to go back to javascript and work on some more practice problems, I got too soft doing backend lol.
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:153]: You got it! Then hit the million combinations! Better to have working code than and refactor than to not. 

Your pseudo code is on point and you are just missing the goal of returning a boolean in your code other than that your almost solid!

File: interview-questions.md [hunk 7]
```diff
@@ -8,44 +8,44 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 4. What is a relational database? Are there other kinds of databases?
 
-Your answer:
+Your answer: A relational database is one that translates one language into something a database can read, then translates it back so that the programmer can read it. 
 
-Researched answer:
+Researched answer: A relational database is an idea that when data is stored, you can use relationships between two things to access it. An example of this is in a username that one might create on a website. The customer creates their username, and it is stored as a primary key, but while they do that a copy is made on the hard drive of the websites server somehwere, stored as a foreign key. Now that there is a relationship between these two things, they can have more data that is stored from that primary key and foreign key relationship. There are other kinds of databases, such as a non relational database.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:31]: This is very close! I really appreciate how well you articulated yourself beacuse it helps me out clearing some stuff up. 

In a table called User we would have:
- a primary key that is different for each user
- a username which would be a string
- a password which would be a string
- and maybe some specific other data

If I wanted my User to be able to make Post's, I would make another Table called Posts. 
I can associatie the tables using model relationships like the User has many posts and each post belongs to a User. 

My post table would have something like: 
- primary key for tracking all the different posts
- a title which is a string
- text which is a string or text
- and a foreign key for identifying who the author / User is that wrote it.

File: interview-questions.md [hunk 8]
```diff
@@ -8,44 +8,44 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 2. What is a gem?
 
-Your answer:
+Your answer:  A gem is something that we install onto ruby in order for it to have more uses than it comes with out of the box. An example of this would be in ruby testing, or rspec. This needs to be installed, but once you have it it is very useful. 
 
-Researched answer:
+Researched answer: A gem is  a package that ruby developers will download in order to add more onto what ruby comes standard with. RubyGems is the package manager that people use to do this. One example of this is rpsec, which allows people to test their code, similarly to how yarn works for javascript. Gems are generally used to modify or add onto what ruby, or ruby applications, can already do.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:19]: Nice!

File: interview-questions.md [hunk 9]
```diff
@@ -8,44 +8,44 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 1. What is a hash in Ruby?
 
-Your answer:
+Your answer: A hash in Ruby is a unique key that relates to an assigned object or class. It is long, and contains letters and numbers in a random order. This is so that many can be assigned and no one has the same hash.
 
-Researched answer:
+Researched answer: In ruby, a hash is something that comes from a class, and is represented by key value pairs. It is kind of like an object in the way that you can use methods to create many isntances of a hash out of a class that you create. For example, if i was to create a class named Book, I could instantiate it using the name of the book, the author, and the genre. I could create a Hash by using Book.new, and giving it a name, genre, and author. I could do this with any number of hashes, any create any number of unique books. The deafult state of a hash is nil, until updated.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:13]: Well said!

File: ruby-challenges.rb [hunk 10]
```diff
@@ -10,23 +10,74 @@
...
+        stopped = 0
+        if @current_speed > 0
+            @current_speed -= brake
+            p @current_speed
+        elsif @current_speed <= 0
+            @current_speed = 0
+         else
+            p "The bike isn't moving"
+        end
+    end
```
Comment (WalkerTexas-coder) [inline ruby-challenges.rb:66]: This is really good! Love the message

**Review by WalkerTexas-coder — APPROVED** [review]
Sam this is really solid! Honestly, I have run into the same blockers tons of times. What's great is that if you moved past these blockers most all of your code works fine! Check out my feedback below-- keep up the good work you are actually on the right path!
